File: String/3. Longest Substring Without Repeating Characters/Solution.java
/*
 时间复杂度 O(n)
 空间复杂度 O(n)

 滑动窗口问题，[l, r] 表示目前的没有重复子串，r向前移动试探下一个字符是否重复，
 采用map存储当前子串中每个字符是否重复

*/


 class Solution {
    public int lengthOfLongestSubstring(String s) {
        int l = 0;
        int r = -1;
        int length = 0;
        HashMap<Character, Boolean> map = new HashMap<Character, Boolean>();
        while(l < s.length()){
            
            if(r+1<s.length() && (!map.containsKey(s.charAt(r+1)) || !map.get(s.charAt(r+1)))){
                
                map.put(s.charAt(++r), true);
            }else{
                map.put(s.charAt(l++), false);
            }
            
            length = Math.max(length, r-l+1);
        }
        
        return length;
        
    }
}

File: String/3. Longest Substring Without Repeating Characters/Solution.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        l = 0
        r = -1
        length = 0
        map = {}
        while l < len(s):
            if r+1 < len(s) and (not(s[r+1] in map) or not(map.get(s[r+1]))):
                r += 1
                map[s[r]]=True
            else:
                map[s[l]]=False
                l += 1
            length = max(length,r-l+1)
        return length